packages/block-library/src/style.scss

```diff
@@ -216,30 +216,27 @@
 
 
 // Font sizes.
-:root {
-	.has-small-font-size {
-		font-size: 13px;
-	}
-
-	.has-regular-font-size, // Not used now, kept because of backward compatibility.
-	.has-normal-font-size {
-		font-size: 16px;
-	}
+.has-small-font-size {
```

Comment:
This is working to fix the issue in #21759, however I'm not sure the bare `.has-xx-font-size` classes are actually specific enough to work out of the box now. When testing with a theme that supplies no editor styles (Gutenberg Starter Theme), it appears that these rules are just overridden by the default `.editor-styles-wrapper p` font size rule. 

<img width="1364" alt="Screen Shot 2020-04-29 at 9 15 47 AM" src="https://user-images.githubusercontent.com/1202812/80600321-37d8b600-89fa-11ea-819f-e91d2284736b.png">


Do we need to at least change these to `.editor-styles-wrapper p` at least? That should allow them to kick in by default, but still be overridden by theme styles (assuming the theme styles have the `.editor-styles-wrapper` selector added via `add_editor_style()`).

Reply:
The root problem seems to me that we bump up the specificity for theme's stylesheets but not for its dependencies (the defaults provided by Gutenberg). The ideal solution would be that we also wrap the block-library stylesheets with `editor-styles-wrapper` at runtime. However, I fear that ship may have already sailed. I'll be more than extremely happy to be proved wrong and go with this approach, though.

Alternative solution: increase the specificity of these classes, only for the editor. My only concern is how would we do it. All the options I see make me sigh. I pushed a fix by adding the `editor-styles-wrapper` class in this file, although I know that we tend to avoid adding the editor wrapper class to anything that goes to the front-end. The alternatives I've considered include:

- Replicate the classes in both files => we enqueue the classes twice + duplicate source of truth (if we change them in the future we'll have to update both places and can be easily forgotten).
- Extract them to a file that will be imported here and in editor.scss => we enqueue the classes twice + it adds indirection.

Note that this happens in every theme that don't provide a font size palette themselves (for default themes: all but TwentyNineteen and TwentyTwenty) since we [no longer inline the style declaration for presets](https://github.com/WordPress/gutenberg/pull/21428).

---

[1] And also TwentyNineteen, but for different reasons: although it does provide a palette, it is identical to Gutenberg's except for the removal of Medium text size, so it doesn't care about enqueuing the classes itself. For that reason, it behaves like the themes that don't opt-in into the font-size palette. It could be argued that this case should be fixed in the theme itself. If we provide a fix for the other themes, we're also fixing this, though.

Reply:
What if we extracted the palettes (color, font, gradients) to its own stylesheets and only enqueue them _if_ the theme hasn't declared support?

It somehow makes extracting the files clearer for me, it'd be reasonable that they contain styles for both editor&front-end, and also is good practice not to ship CSS we don't use. At the same time, it is also a bit more work than what we have here and, given the sheer amount of things to do, I find it difficult to justify (although it's the approach I'd do otherwise).

Reply:
> Alternative solution: increase the specificity of these classes, only for the editor. My only concern is how would we do it. All the options I see make me sigh. I pushed a fix by adding the editor-styles-wrapper class in this file, although I know that we tend to avoid adding the editor wrapper class to anything that goes to the front-end. 

Can't we just have two sets of styles — one in `editor.scss`, and one in `theme.scss`? It would be in two separate places, but that's what we usually do when we need separate styles for the front and back end.

Reply:
Oh, but isn't `theme.scss` only enqueued if theme declares support for `wp-block-styles` (which is something themes that don't declare font-size palettes are less likely to do)? I thought that wouldn't fix the issue for most themes.

Reply:
Whoops, nevermind — you are right. I'm not sure what our best solution is then. 😞 

cc-ing @jasmussen for a set of fresh eyes.

Reply:
Thanks for the ping. Trying to catch up, but this is gnarly. 

My first instinct was to look, what do we do with colors? I was certain we used to do `.has-vivid-red-color.has-vivid-red-color` to ensure specificity. But now it appears we do this:

```
:root .has-vivid-red-color {
    color: rgb(207, 46, 46);
}
```

So color does what we appear to be reverting here, and to me the two feel related. Do we make a similar change for colors, or is that different, and if so, why is it different?

---

Going back to the topic at hand, this PR does seem to work for me:

<img width="2094" alt="Screenshot 2020-05-01 at 09 04 15" src="https://user-images.githubusercontent.com/1204802/80788923-c714d380-8b8a-11ea-9164-863419a52b12.png">

The dual inclusion of `.has-xx-font-size` and `.editor-styles-wrapper .has-xx-font-size` correctly overrides the p font size rule I've set in my theme:

<img width="2094" alt="Screenshot 2020-05-01 at 09 04 15" src="https://user-images.githubusercontent.com/1204802/80789057-2ffc4b80-8b8b-11ea-9b74-d88c35baf59e.png">

Reply:
😄 This is really confusing, but I think I can outline the whole situation. 

## The current state of color + font size options CSS 

Prior to this PR, the default colors _and_ font styles include a `:root` prefix to both the front and back end. For example: 

```
:root .has-huge-font-size { font-size: 72px }
:root .has-vivid-red-color { color: red }
```

This works great in the editor. But since themes generally provide their custom overrides without a prefix like that, these rules take precedence over the custom theme rules in the front end: 

```
:root .has-huge-font-size { font-size: 72px } // Default Gutenberg Rule
.has-huge-font-size { font-size: 72px } // Custom Theme Rule

:root .has-vivid-red-color { color: red } // Default Gutenberg Rule
 .has-vivid-red-color { color: blue } // Custom Theme Rule
```

That means that the custom theme rule will be ignored completely in the front end. 😞 

This is only an issue if a theme provides the exact same name as Gutenberg's default for their custom sizes and colors. That's a frequent issue with custom sizes, but less of an issue for custom colors (but it should still be fixed). 


## This PR

This PR removes that `:root` selector from the font size rule, to allow custom theme font styles to work again on the front end: 

```
.has-huge-font-size { font-size: 72px } // Custom Theme Rule
.has-huge-font-size { font-size: 72px } // Default Gutenberg Rule
```

As @aduth noted [in the original ticket](https://github.com/WordPress/gutenberg/issues/21759#issuecomment-621169176), this relies on load order of the CSS, which isn't excellent. But it works.

In making this change however, this PR ran into another problem: Gutenberg [includes a](https://github.com/WordPress/gutenberg/blob/c19d2d908cba695960cf8407bd0b0afc181aa657/packages/editor/src/editor-styles.scss#L97) `p { font-size: inherit; }` rule that is prefixed with `.editor-styles-wrapper`. This has higher specificity than the new `.has-xx-font-size`, so it cancels out the custom font size rule in the editor. 

The new solution @nosolosw tried in https://github.com/WordPress/gutenberg/pull/21969/commits/163143a00418a163c38db14fe0e6129b78bea0a7 adds an additional `.editor-styles-wrapper .has-xx-font-size` rule to get around that.  This works, but it provides an unnecessary `.editor-styles-wrapper .has-xx-font-size` rule to the frontend. Not a huge problem, but not ideal. 

Whatever we do here, we'll want to make the same change to the custom colors rules, just in the off chance that a theme has specified a custom color that uses one of Gutenberg's default color names. Currently, the theme's front-end color style would be overridden in that case.

Reply:
Thank you for explaining so thoroughly, Kjell.

Master:

> In other words, the custom theme rule will be ignored completely in the front end. 😞

So far, so good.

This PR:

> .has-huge-font-size { font-size: 72px } // Custom Theme Rule
> .has-huge-font-size { font-size: 72px } // Default Gutenberg Rule

Cool, works because it comes after, also good. Side question, though: when a theme referenced class name such as this one clashes with Gutenberg, is either the theme or the editor doing something wrong? 

> p { font-size: inherit; } rule that is prefixed with .editor-styles-wrapper. This has higher specificity than the new .has-xx-font-size, so it cancels out the custom font size rule in the editor.

Yes, I'm running into this with my own theme I'm building, where this:

```
p {
	margin-top: 1em;
	margin-bottom: 1em;
}
```

in the editor is cancelled out by:

```
.editor-styles-wrapper .block-editor-block-list__block {
	margin-top: 28px;
	margin-bottom: 28px;
}
```

It seems a similar issue, the tension between the editor wanting to provide some baseline styles, and the theme needing to trivially override those. 

> This works, but it provides an unnecessary .editor-styles-wrapper .has-xx-font-size rule to the frontend. Not a huge problem, but not ideal.

There we go, that seems to be the crux of the issue. 

As I read this, my instinct is to think that if we provide a thorough comment, this is an okay tradeoff for now. Take [the entire introduction in the normalization file](https://github.com/WordPress/gutenberg/blob/c19d2d908cba695960cf8407bd0b0afc181aa657/packages/editor/src/editor-styles.scss#L7), it's one big "ugh wp-admin bleed" disclaimer. So it all trails back to that.

But we are more aware than ever! An iframe is potentially on the horizon, which would allow us to throw out most, if not all, of the normalization stylesheet, prompting us to be able to refactor this also.

I wonder if we should make a specific keyword for such a refactor — in my own private projects, I often write `@todo` so I can search for that identifier. We could make a `@todo-editor-styles`, which we can then search for if either an iframe, or Shadow DOM protection, or better scoping of wp-admin styles land.

Reply:
> Side question, though: when a theme referenced class name such as this one clashes with Gutenberg, is either the theme or the editor doing something wrong?

That's a good philosophical question. 😄 I don't think either is doing it wrong — we expect some Gutenberg styles to be overridden by the themes in general. 

> As I read this, my instinct is to think that if we provide a thorough comment, this is an okay tradeoff for now. Take the entire introduction in the normalization file, it's one big "ugh wp-admin bleed" disclaimer. So it all trails back to that.

Good point. I think you're probably right. This works for now, and things will be (hopefully) changing soon. So that may be good enough for the moment. @nosolosw what do you think?

Reply:
Reading @kjellr outline above it seems to me that the issue in twofold: how to make palettes work with 1) themes that don't provide one and 2) themes that provide one with the same class names that Gutenberg defaults.

In its current state, this PR fixes 1. In addition to this, should we also enqueue the palettes _only_ if the theme hasn't provided one? That should fix 2 as well.

Sorry if I missed anything and this doesn't make any sense 😅 I'm from mobile and AFK so my brain is in low-power mode. I join the wishes that we can remove this altogether in the not so distant future!

Reply:
Re-reading this thread I gather we're ok with this solution _for now_. Is that right @kjellr @jasmussen ? If so I'd need an :green_heart:

I was thinking this could land shorty so it can be in for next week release. I'd like to avoid having the `:root` for font-sizes for much longer as it was only added for 7.9 because we forgot to remove it after some experiments we did (can be considered a regression).

If the other issue I mentioned (conflicts with themes that provide the same class names than Gutenberg) is pressing I can prepare a different PR to solve that (I believe the answer to that issue would be to not enqueue the color & font-sizes classes if the theme has declared its own).

Reply:
> If the other issue I mentioned (conflicts with themes that provide the same class names than Gutenberg) is pressing I can prepare a different PR to solve that (I believe the answer to that issue would be to not enqueue the color & font-sizes classes if the theme has declared its own).

That sounds like a promising solution. 👍